### PR TITLE
Dj fields

### PR DESCRIPTION
**Pull request by Jacob8399**
dj_fields is a very large merge and comes from a branch that has been active for nearly a year. It implements the option to have multiple different fields in one simulation. Each field has a specific soil profile, crop rotation, and field management. This can be visualized in input/barnyard.json and input/animal_management.json. Each field will be split up in the output files but are separate from feed and pens.

field_managment.py was added to manage tillage, fertilizer, and manure applications. These applications are specified in the field_management json files and can be either scheduled or optimal. Optimal applications are specified in a similar way to crop rotations, app_years are the years you want the application and repeat creates a repeating pattern of 'X' years. 
For example:
        "app_years": [2009],
        "repeat": 1,
This will apply in 2009 and every year after. The optimal application will happen when it is most useful, manure and fertilizer right before planting and tillage after a full harvest. Scheduled applications will have a specific date for applying. Examples of both options can be seen in input/field_management/barnyard_field_management.json.

Many of the other changes are small bugs fixes, optimizations, and commenting changes such as:
Updated each crop type with more accurate default values
Output structure to work with fields
crop and soil folders are now stored in a field folder in routines
Many variable names changed to get away from camel case and to the underscore style names for Python

The initial merge from master to dj_fields had an overwhelming amount of conflicts. The areas of the model I know well should be fine but while reviewing please give extra attention to feed.py and the animal code if you know those files well. 

If there are any questions feel free to email me or message me on slack.

Edit: This branch will increase the simulation time the more fields you add and the more output diagnostics you request. Switching to only one field and 'standard_output.json' in the output section of the main json files will decrease the run time significantly.

### CODE REVIEW

File: RUFAS/output_handler/output_handler.py [hunk 1]
```diff
@@ -41,13 +41,13 @@ def __init__(self, data, state):
 
         # Instantiate Report Handler Objects here
         self.reports = {
-                        'field_report': FieldReport(data['field_report']),
                         'feed_storage_report': FeedStorageReport(data['feed_storage_report']),
-                        'mass_balance_report': MassBalanceReport(data['mass_balance_report']),
```
Comment (militsasotirova) [inline RUFAS/output_handler/output_handler.py:46]: Out of curiosity, why was the mass balance report object taken out here?

Comment (wmaxdonovan) [reply RUFAS/output_handler/output_handler.py:46]: Actually just a very good catch! All of the current mass balance reports operate on the field level, so we reorganized the class structure to include a BaseFieldMassBalance report and a BaseMassBalance report. The driver for the BaseMassBalance is currently empty, but we want to keep it around, so I just forgot to add the empty report back.

File: RUFAS/routines/field/field.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        self.soil_data = read_json_file(input_dir / 'soil' / field_data['soil'])
+        self.crop_data = read_json_file(input_dir / 'crop' / field_data['crop'])
+        self.field_management_data = read_json_file(input_dir / 'field_management' / field_data['field_management'])
+
+        self.soil = Soil(self.soil_data)
+        self.field_management = FieldManagement(self.field_management_data, time)
+        self.crop = Crop(self.crop_data, time)
+
+
+def read_json_file(file_path: Path):
```
Comment (militsasotirova) [inline RUFAS/routines/field/field.py]: read_json_file() is separately defined in 3 different files (field.py, classes.py, and simulation_engine.py). The code for the two in field.py and classes.py seem to be identical, and the one in simulation_engine.py is slightly different, but operates on the same idea - is there any reason why there isn't a single method in (for example) util.py so that we don't have repeated code?

File: RUFAS/routines/field/field_management/field_management.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            """
+            Description:
+                Moves field_management to the next day– called from check_conditions.
+            Args:
+                weather: an instance of the Weather class specified in classes.py
+                    contains information about the environment
+                time: an instance of the Time class specified in classes.py
+            """
+            day = time.day
+            calendar_year = time.start_year + time.year - 1
```
Comment (militsasotirova) [inline RUFAS/routines/field/field_management/field_management.py]: Doesn't time have a calendar_year attribute?

File: RUFAS/routines/field/field.py (hunk 2 above)
Comment (Jacob8399) [reply RUFAS/routines/field/field.py]: Those can definitely be packaged into one method in util, good call.

File: RUFAS/routines/field/field_management/field_management.py (hunk 3 above)
Comment (Jacob8399) [reply RUFAS/routines/field/field_management/field_management.py]: Another good call

File: RUFAS/test/test_handler.py [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
 
 
 def run_tests():
     """
     Calls the unit test functions for each routine that has them.
     """
-    # animal.run_tests()
-    # print('Tests completed.\n')
-    pass
+
```
Comment (militsasotirova) [inline RUFAS/test/test_handler.py:17]: I appreciate the file clean up that's in this branch. I've been thinking about the unit tests lately, and it seems like they aren't relevant anymore, especially with the ration formulation changes that are currently being made (and we aren't using them for any sort of whole-model check). Any thoughts on deleting the associated files/methods? We could always bring them back if we find that it makes sense to have them again.

Comment (wmaxdonovan) [reply RUFAS/test/test_handler.py:17]: Adding unit tests for different components of the model has always been on my todo, so I am in favor of keeping them, but I think you're right that these are becoming increasingly underutilized (especially because I do most of my testing as gut checks on full model runs) @Jacob8399 ?

File: RUFAS/routines/field/field.py (hunk 2 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/field.py]: @Jacob8399 Do you want to handle this or should I?

File: RUFAS/routines/field/field_management/field_management.py (hunk 3 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/field_management/field_management.py]: Same thing for references to field_management objects. I make the change in dm_manure, so it should not be necessary here, but if you see it later in the review, know that it has been updated in later branches

File: RUFAS/test/test_handler.py (hunk 4 above)
Comment (Jacob8399) [reply RUFAS/test/test_handler.py:17]: I agree with Max. I think they could be useful and we should at least leave the structure in place. The unit tests have never been prioritized on our end, even though they probably should be.

File: RUFAS/routines/field/field.py (hunk 2 above)
Comment (Jacob8399) [reply RUFAS/routines/field/field.py]: Already did it

File: RUFAS/test/test_handler.py (hunk 4 above)
Comment (militsasotirova) [reply RUFAS/test/test_handler.py:17]: If we have plans/ideas for unit tests for the rest of the model, then I agree we should keep them (and edit the ones based on the old ration). We could discuss this at our next programmer's meeting? (I do think that there is a lot of value in testing - I just want to make sure we have a good idea of what we want to do with it.) We could maybe start a branch solely dedicated to developing the unit test folder? (And I also think that if we decide to move forward with this, we should do some planning around it - e.g. what edge cases are relevant, what are some "normal" cases to test, etc.)

Comment (Jacob8399) [reply RUFAS/test/test_handler.py:17]: Discussing it at the next programmer's meeting sounds good. I think we need a better idea for what actually qualifies as a unit test or not.

Comment (militsasotirova) [reply RUFAS/test/test_handler.py:17]: [This](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5007758/) article discusses distinctions between different kinds of tests (warning - the code snippets are formatted horribly, at least on my browser). The Conclusion is a good summary.

**Comment by militsasotirova** [comment]
What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?

**Comment by militsasotirova** [comment]
> What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?

Actually nevermind I think its the output.

**Comment by Jacob8399** [comment]
> What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?

I'm not actually sure. In what situation did this really show? I did a quick test and the ratios were similar so it might depend on computer speed during the simulation too.

**Comment by militsasotirova** [comment]
> > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> 
> I'm not actually sure. In what situation did this really show? I did a quick test and the ratios were similar so it might depend on computer speed during the simulation too.

So it took 127 seconds to run barnyard.json in this branch, so I went back to master where it took 82. Then I came back to this one and it took 75? So I don't really know (I probably mixed something up honestly). Should we have the simulation time just be the model run time without output generation? It would be easier to spot differences in time.

**Comment by wmaxdonovan** [comment]
> > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> 
> Actually nevermind I think its the output.



> > > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> > 
> > 
> > I'm not actually sure. In what situation did this really show? I did a quick test and the ratios were similar so it might depend on computer speed during the simulation too.
> 
> So it took 127 seconds to run barnyard.json in this branch, so I went back to master where it took 82. Then I came back to this one and it took 75? So I don't really know (I probably mixed something up honestly). Should we have the simulation time just be the model run time without output generation? It would be easier to spot differences in time.

The other thing is that barnyard may be set to run multiple fields. That should scale runtime linearly (theoretically I guess)

**Comment by Jacob8399** [comment]
> > > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> > 
> > 
> > Actually nevermind I think its the output.
> 
> > > > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> > > 
> > > 
> > > I'm not actually sure. In what situation did this really show? I did a quick test and the ratios were similar so it might depend on computer speed during the simulation too.
> > 
> > 
> > So it took 127 seconds to run barnyard.json in this branch, so I went back to master where it took 82. Then I came back to this one and it took 75? So I don't really know (I probably mixed something up honestly). Should we have the simulation time just be the model run time without output generation? It would be easier to spot differences in time.
> 
> The other thing is that barnyard may be set to run multiple fields. That should scale runtime linearly (theoretically I guess)

I think those discrepancies are just computer running speed at those moments. I can run barnyard back to back and get 18 seconds and then 25 seconds (with standard output and one field).

**Comment by militsasotirova** [comment]
> > > > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> > > 
> > > 
> > > Actually nevermind I think its the output.
> > 
> > 
> > > > > What's causing the jump in computation time for barnyard.json (while animal_management.json takes the ~same amount of time)?
> > > > 
> > > > 
> > > > I'm not actually sure. In what situation did this really show? I did a quick test and the ratios were similar so it might depend on computer speed during the simulation too.
> > > 
> > > 
> > > So it took 127 seconds to run barnyard.json in this branch, so I went back to master where it took 82. Then I came back to this one and it took 75? So I don't really know (I probably mixed something up honestly). Should we have the simulation time just be the model run time without output generation? It would be easier to spot differences in time.
> > 
> > 
> > The other thing is that barnyard may be set to run multiple fields. That should scale runtime linearly (theoretically I guess)
> 
> I think those discrepancies are just computer running speed at those moments. I can run barnyard back to back and get 18 seconds and then 25 seconds (with standard output and one field).

Ah ok - thanks!

**Review by militsasotirova — APPROVED** [review]

File: RUFAS/routines/animal/animal_management.py [hunk 5]
```diff
@@ -447,6 +461,7 @@ def calc_manure_excretion(self, feed):
         Args:
```
Comment (HectorMM3) [inline RUFAS/routines/animal/animal_management.py:461]: Jacob.  Is "Args" defined somewhere? I see descriptions but I don't know if this is a shortened word or abbreviation...

File: RUFAS/routines/field/crop/crop_types/alfalfa.py [hunk 6]
```diff
@@ -97,16 +100,18 @@ def __init__(self, data):
...
+        self.epco = 1
 
         self.water_actual_up = 0
         self.water_uptake_each_layer = []
 
         # ===================================================================
         ''' Nitrogen Uptake Data '''
 
-        self.beta_n = 10  # TODO: value taken from corn, value for alfalfa unknown
+        self.N_fix = 0.0
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop_types/alfalfa.py:112]: Where is Nfix in the pseudocode?

Comment (HectorMM3) [reply RUFAS/routines/field/crop/crop_types/alfalfa.py:112]: What does  self.N_fix = 0.0 mean?

File: RUFAS/routines/animal/animal_management.py (hunk 5 above)
Comment (Jacob8399) [reply RUFAS/routines/animal/animal_management.py:461]: Args stands for arguments which are what we call variables passed into methods.

File: RUFAS/routines/field/soil/phosphorus_cycling/manure_leach.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    day = time.day
+    year = time.year
+    rainfall = weather.rainfall[year - 1][day - 1]
+    runoff = soil.runoff
+    temp = weather.T_avg[year - 1][day - 1]
+
+    # Calculate manure factors
+    # S.5.G.I
+
+    # S.5.G.I.1
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/phosphorus_cycling/manure_leach.py:39]: I could not find S.5.G.I in the soil pseudocode.

File: RUFAS/test/test_handler.py (hunk 4 above)
Comment (HectorMM3) [reply RUFAS/test/test_handler.py:17]: We should definitely have someone check out units. Unit consistency is a big deal in mathematical models.

File: save_directory/CSVs/field_report/soil_report/soil_report.csv [hunk 8]
```diff
@@ -1,5 +1,5 @@
 year,j_day,precip,runoff,ET_max,ET_act,trans_max,evap_max,surface_temp,sediment_yield,residue,trans_act_L1,trans_act_L2,trans_act_L3,soil_water_L1,soil_water_L2,soil_water_L3,evap_L1,evap_L2,evap_L3,perc_L1,perc_L2,perc_L3,temperature_L1,temperature_L2,temperature_L3
-,,mm,mm,mm d^-1,mm H20,mm H20,mm H20,C,metric tons,kg/ha,mm H20,mm H20,mm H20,mm,mm,mm,mm H20,mm H20,mm H20,mm H20,mm H20,mm H20,C,C,C
```
Comment (HectorMM3) [inline save_directory/CSVs/field_report/soil_report/soil_report.csv:2]: What do the "C,C,C" stand for at the end?

**Review by wmaxdonovan — APPROVED** [review]
This branch is simulating multiple fields correctly. There are a number of changes made to application management in dm_manure, and we have some ideas for aggregating output at different levels of the model. I think this branch is ready to be merged so we can all start working with the multiple fields representation.

**Review by HectorMM3 — CHANGES_REQUESTED** [review]

File: RUFAS/routines/animal/animal_management.py (hunk 5 above)
Comment (HectorMM3) [reply RUFAS/routines/animal/animal_management.py:461]: Solid. Thank you for the clarification.

File: save_directory/CSVs/field_report/soil_report/soil_report.csv (hunk 8 above)
Comment (wmaxdonovan) [reply save_directory/CSVs/field_report/soil_report/soil_report.csv:2]: Degrees celsius

File: RUFAS/routines/field/crop/crop_types/alfalfa.py (hunk 6 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/crop/crop_types/alfalfa.py:112]: N_fix is fixed Nitrogen, it is pseudocode_crop [C.5.D.1]. The pseudocode is not referenced here because this is an object assignment. self.N_fix = 0.0 just means "there is a variable in this object called N_fix that will be some floating point number". Self is always in reference to the object that the statement is contained within. In this case, Alfalfa

File: RUFAS/routines/field/soil/phosphorus_cycling/manure_leach.py (hunk 7 above)
Comment (wmaxdonovan) [reply RUFAS/routines/field/soil/phosphorus_cycling/manure_leach.py:39]: The indexes in the pseudocode were updated to reflect the transfer of the manure, tillage, and fertilizer components of phosphorus cycling to the field_management section of the code in dm_manure. The equations you are looking for are under S.5.D.I.

**Review by HectorMM3 — APPROVED** [review]
Jacob, I have reviewed each line of code including changes and deleted code. The changes seem appropriate and well documented. I have outlined some potential areas for model failure, but I believe these need to be addressed on a soil and crop level not a merge level. As for the merge, the structure of the model appears to be addressed in order to bring this into Master.  I will email my outline in the Slack Channel or on our meeting Thursday.

**Review by tangswm — APPROVED** [review]
I've gone over all the changes and ran a bunch of simulations and dj_fields seems to be working well. There is more functionality that I'm working on for management, but as of right now the branch seems ready to merge.

File: RUFAS/routines/feed/feed.py [hunk 9]
```diff
@@ -15,50 +20,75 @@
...
+            # search for matching storage profile
+            for storage in feed.available_storage.values():
+                if storage.feed_id == crop.feed_id and storage.storage is True \
+                        and storage.storage_quality == crop.harvest_quality:
+                    storage.store_crop(crop)
+                    stored = True
+
+            # search for available, empty storage
+            if not stored:
+                for storage in feed.available_storage.values():
```
Comment (cvankerkhove) [inline RUFAS/routines/feed/feed.py:55]: If this boolean statement is reached and there are multiple storage receptacles in feed.available_storage.values() with DM==0 would this for loop store the crop multiple times in different storage objects?

**Review by cvankerkhove — APPROVED** [review]
Overall the new fields structure looks good and integrates well with the current feed management system. I made some minor changes in the most recent commit to make the multiple fields cases run smoothly with crop storage and management. I also ran multiple scenarios with simulating different crops and it worked well.

File: RUFAS/routines/feed/feed.py (hunk 9 above)
Comment (Jacob8399) [reply RUFAS/routines/feed/feed.py:55]: I don't believe so because stored will be set to true at the bottom of the loop and then the inner if statement won't be true. @wmaxdonovan wrote this code so he can double-check that logic.

Comment (Jacob8399) [reply RUFAS/routines/feed/feed.py:55]: Actually I just checked your changes and it seems you added the "and not stored" which should fix it.

Comment (wmaxdonovan) [reply RUFAS/routines/feed/feed.py:55]: @Jacob8399 @cvankerkhove Feed management for multiple fields with similar and different crops seems to be working as intended. boolean stored is reset to false for each crop, set to true after any of the three storage scenarios, and each storage scenario can only be entered if stored is false. Happy to see similar crops ending up in the same silo, I was worried about that.
